ironfish/src/memPool/memPool.ts

```diff
@@ -239,18 +270,79 @@ export class MemPool {
 
     this.transactions.set(hash, transaction)
 
-    this.transactionsBytes += getTransactionSize(transaction)
-
     for (const spend of transaction.spends) {
-      if (!this.nullifiers.has(spend.nullifier)) {
-        this.nullifiers.set(spend.nullifier, hash)
-      }
+      this.nullifiers.set(spend.nullifier, hash)
     }
 
     this.queue.add({ hash, feeRate: getFeeRate(transaction) })
-    this.expirationQueue.add({ expiration: transaction.expiration(), hash })
+    this.evictionQueue.add({ hash, feeRate: getFeeRate(transaction) })
+    this.expirationQueue.add({ hash, expiration: transaction.expiration() })
+
+    this.transactionsBytes += getTransactionSize(transaction)
     this.metrics.memPoolSize.value = this.count()
-    return true
+
+    // TODO: Add telemetry for evictions
+    if (this.full()) {
+      this.evictTransactions()
+    }
+
+    return this.transactions.has(hash)
+  }
+
+  /**
+   * @returns true if the mempool is over capacity
+   */
+  full(): boolean {
+    return this.sizeBytes() >= this.maxSizeBytes
+  }
+
+  /**
+   * @returns true if a transaction hash is in the recently evicted cache
+   */
+  recentlyEvicted(hash: TransactionHash): boolean {
+    return this.recentlyEvictedCache.has(hash.toString('hex'))
+  }
+
+  /**
+   * Evicts transactions from the mempool until it is under capacity.
+   * Transactions are evicted in order of fee rate (descending).
+   *
+   * @returns an array of the evicted transactions
+   */
+  private evictTransactions(): Transaction[] {
+    const evictedTransactions: Transaction[] = []
+
+    while (this.full()) {
+      const next = this.evictionQueue.peek()
+
+      const transaction = next && this.transactions.get(next.hash)
+      if (!transaction) {
+        break
```

Comment:
```suggestion
        // need to remove the txn from the eviction queue
        continue
```
1. imo continue > break if the transaction cannot be found for some reason we should continue to evict
2. need to remove this faulty txn from the eviction queue as otherwise on the next call of `evictTransactions` nothing will be evicted as `queue.peek()` will return the same obj at the front of the pq

Reply:
Discussed, we'll change this to `Assert` that transaction is always defined. This should never be undefined because if the mempool is full there should always be another transaction present